### ISSUE TITLE
Training in batches in TF Slim

### ISSUE THREAD

**Issue by deepankar1994** [tensorflow/models#956]
I am largely following [this]https://github.com/tensorflow/models/blob/master/slim/slim_walkthough.ipynb).

```
# Note that this may take several minutes.

import os

from datasets import flowers
from nets import inception
from preprocessing import inception_preprocessing

slim = tf.contrib.slim
image_size = inception.inception_v1.default_image_size


def get_init_fn():
    """Returns a function run by the chief worker to warm-start the training."""
    checkpoint_exclude_scopes=["InceptionV1/Logits", "InceptionV1/AuxLogits"]
    
    exclusions = [scope.strip() for scope in checkpoint_exclude_scopes]

    variables_to_restore = []
    for var in slim.get_model_variables():
        excluded = False
        for exclusion in exclusions:
            if var.op.name.startswith(exclusion):
                excluded = True
                break
        if not excluded:
            variables_to_restore.append(var)

    return slim.assign_from_checkpoint_fn(
      os.path.join(checkpoints_dir, 'inception_v1.ckpt'),
      variables_to_restore)


train_dir = '/tmp/inception_finetuned/'

with tf.Graph().as_default():
    tf.logging.set_verbosity(tf.logging.INFO)
    
    dataset = flowers.get_split('train', flowers_data_dir)
    images, _, labels = load_batch(dataset, height=image_size, width=image_size)
    
    # Create the model, use the default arg scope to configure the batch norm parameters.
    with slim.arg_scope(inception.inception_v1_arg_scope()):
        logits, _ = inception.inception_v1(images, num_classes=dataset.num_classes, is_training=True)
        
    # Specify the loss function:
    one_hot_labels = slim.one_hot_encoding(labels, dataset.num_classes)
    slim.losses.softmax_cross_entropy(logits, one_hot_labels)
    total_loss = slim.losses.get_total_loss()

    # Create some summaries to visualize the training process:
    tf.scalar_summary('losses/Total Loss', total_loss)
  
    # Specify the optimizer and create the train op:
    optimizer = tf.train.AdamOptimizer(learning_rate=0.01)
    train_op = slim.learning.create_train_op(total_loss, optimizer)
    
    # Run the training:
    final_loss = slim.learning.train(
        train_op,
        logdir=train_dir,
        init_fn=get_init_fn(),
        number_of_steps=2)
        
  
print('Finished training. Last batch loss %f' % final_loss)
```
My issue is that it seems that the training is being done on the same subset of images on every iteration i.e. only one call to the load_batch function seems to be made, what is the proper way to get it to train using a different set of image batch during every training step ?

**Comment by drpngx** [comment]
You need to use a random shuffle. You can do that with

* a random shuffle queue
* a placeholder and randomize before each step
* randomized indices in tensorflow

For the last one, here's what I have for mnist

```
  train_indices = tf.random_uniform(
      [BATCH_SIZE], 0, num_train_samples, tf.int64)
  train_data_node = tf.gather(tf_train_data, train_indices)
  train_labels_node = tf.gather(tf_train_labels, train_indices)
```

In the future, this is a question best answered on stackoverflow, since it is a user issue.

**Comment by kindlychung** [comment]
Still no idea how to integrate `train_data_node` into the OP's code. Could you elaborate more? Thanks! @drpngx

**Comment by kindlychung** [comment]
Could you share your mnist code? Thanks! @drpngx
